### ISSUE TITLE
Properties with option read = false to do not appear in query results

### ISSUE THREAD

**Issue by bartylla** [sebelga/gstore-node#96]
I updated from gstore-node version 2.1.4 to 4.0.0. As a result I have a problem with all schemas which use the option read=false in the property definitions. This properties simply do not appear any longer in the query results (or better they are null or only the default value). In general I expected that I'm able to fix this by using the plain version via `entity.plain({ readAll: true });` but also this was not successfully.

Unfortunately the release notes do not mention this change. What I have to do to get the data back in the query results? Or is it simply a bug?  

**Schema:**
```
const userSchema = new Schema({
    email: {
        type: 'string',
        validate: 'isEmail',
        required: true
    },
    password: {
        type: 'string',
        validate: {
            rule: 'isLength',
            args: [{ min: 8, max: undefined }]
        },
        required: true,
        read: false,
        excludeFromIndexes: true
    },
    state: {
        type: 'string',
        default: 'requested',
        write: false,
        read: false,
        excludeFromIndexes: true
    }
});
```

**Query Result**
```
{ 
  id: '5652383656837120',
  email: 'test@test.com',
  state: 'requested',
  password: null,
  ...
}
```

**Plain Result**
```
{ 
  id: '5652383656837120',
  email: 'test@test.com',
  state: 'requested',
  password: null,
  ...
}
```

**Comment by sebelga** [comment]
Hi,
This is strange. I am going to have a look, thanks for reporting 👍

**Comment by sebelga** [comment]
I just made a test and the "read" property works as expected. This is the test I am running:

```js
const schema = new Schema({
    email: {
        type: 'string',
        validate: 'isEmail',
        required: true,
    },
    password: {
        type: 'string',
        validate: {
            rule: 'isLength',
            args: [{ min: 8, max: undefined }],
        },
        required: true,
        read: false,
        excludeFromIndexes: true,
    },
    state: {
        type: 'string',
        default: 'requested',
        write: false,
        read: false,
        excludeFromIndexes: true,
    },
});

const User = gstore.model('User', schema);
const user = new User({ email: 'test@test.com', password: 'abcd1234' });

user.save().then((entity) => {
    const response = entity.plain();
    expect(response.password).to.not.exist;
    expect(response.requested).to.not.exist;

    const response2 = entity.plain({ readAll: true });
    expect(response2.password).equal('abcd1234');
    expect(response2.state).equal('requested');
});
```

Could you give me your scenario where it does not work?

**Comment by sebelga** [comment]
I am closing this for no activity. Feel free to reopen if needed.

**Comment by sebelga** [comment]
Hello,
What you showed is the correct behavior. I can see that the `read: false` does not return "createdOn", "modifiedOn" and "password". You seem to have a different issue. Your middleware does not attach the password to the entity.
Can you provide the full `hashPassword()` method? thanks

**Comment by joelvoss** [comment]
Thank you for the fast response! Here is the complete hashPassword function:

```Javascript
async function hashPassword() {
  const password = this.password;

  // Short circuit, if no password field is present.
  if (password === null || typeof password === 'undefined') {
    return;
  }

  // Hash the password
  try {
    const salt = await bcrypt.genSalt(5);
    const hash = await bcrypt.hash(this.password, salt);
    this.password = hash;
    return;
  } catch (err) {
    throw new Error(err);
  }
}
```

If the password is configured as "read=true", it is returned for all queries with the hashed value.
I also tried the hashPassword function from your example, which had the same result unfortunately.

As a side note: I also use your gstore-api lib, but that shouldn't have anything to do with this problem, should it?

**Comment by sebelga** [comment]
This part here

```js
if (password === null || typeof password === 'undefined') {
    return;
}
```
will set a `null` password.   
What is it exactly that you try to achieve?

If you want to validate that the data contains a password (with `required: true` on your Schema) you should call `validate()` on the entity before trying to save it.
https://sebelga.gitbooks.io/gstore-node/content/entity/methods/validate.html

example:

```js
const user = new User({ email: 'abc@abc.com' }); // password missing
const { error } = user.validate();

if (error) {
// --->  handle error
}
```

**Comment by joelvoss** [comment]
Even if i remove the if statement, the error still occurs :(
I am logging the hash value inside the hashPassword function and it definitely get's generated here and assigned to this.password.

If `read=true` is defined inside the schema, everything works fine. As soon as read=false is set, the hash value is `null`. I will set up your example from the gstore-api repo with the new gstore-node version and test it again, because it more or less resembles what i do in my code.

**Comment by joelvoss** [comment]
Okay, after digging around in the gstore-node sources, i figured it out.

Essentially, I tried to authenticate a user via username/password.
For this I created a `/login` route, which first searches for the correct entity using `.findOne()`. The `.findOne()` call was executed without the option `readAll:true`.

For some reason, I expected the returned entity to be a "untouched" entity, which obviously is not the case. After that i called the `verifyPassword` function of the schema with this entity, which at this point had its password field set to `null`.

So the mistake was mine. However, the example from the gstore-api repo should be revised at this point, since the example also makes the same error there.

Thanks for the help! Really appreciate your libraries.
